File: packages/ios/Sources/CrumbUI/CrumbQualityInstrumentation.swift
import Foundation

public enum CrumbQualityEventKind: String, Sendable {
    case formReady
    case diagnosticsReady
    case screenshotReady
    case reporterClosed
}

public struct CrumbQualityEvent: Sendable {
    public let kind: CrumbQualityEventKind
    public let elapsedMilliseconds: Double
}

@MainActor
public enum CrumbQualityInstrumentation {
    private static var observer: ((CrumbQualityEvent) -> Void)?

    /// Intended for Crumb's release harness, not application analytics.
    @_spi(Quality)
    public static func observe(_ observer: ((CrumbQualityEvent) -> Void)?) {
        self.observer = observer
    }

    static func record(kind: CrumbQualityEventKind, startedAtNanoseconds: UInt64) {
        guard let observer else { return }
        let elapsed = Double(DispatchTime.now().uptimeNanoseconds - startedAtNanoseconds) / 1_000_000
        observer(CrumbQualityEvent(kind: kind, elapsedMilliseconds: elapsed))
    }
}

File: packages/android/crumb-ui/src/main/kotlin/dev/crumb/ui/CrumbQualityInstrumentation.kt
package dev.crumb.ui

/** Release-QA timing boundaries. Dormant unless a host explicitly installs an observer. */
enum class CrumbQualityEventKind {
    FORM_READY,
    DIAGNOSTICS_READY,
    SCREENSHOT_READY,
    REPORTER_CLOSED,
}

data class CrumbQualityEvent(
    val kind: CrumbQualityEventKind,
    val elapsedMilliseconds: Double,
)

object CrumbQualityInstrumentation {
    @Volatile
    private var observer: ((CrumbQualityEvent) -> Unit)? = null

    /** Intended for Crumb's release harness, not application analytics. */
    @JvmSynthetic
    fun observe(observer: ((CrumbQualityEvent) -> Unit)?) {
        this.observer = observer
    }

    internal fun record(
        kind: CrumbQualityEventKind,
        startedAtNanos: Long,
    ) {
        val activeObserver = observer ?: return
        val elapsed = (android.os.SystemClock.elapsedRealtimeNanos() - startedAtNanos) / 1_000_000.0
        activeObserver(CrumbQualityEvent(kind, elapsed))
    }
}
